rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,184 @@
+        // the last entry for incomplete sigs and add the current send to account for multiple methods chained on sig
+        //
+        // E.g.: `sig.override.final {}`
+        //
+        // If we never find a send where the first receiver is a `sig` and the last invocation has the declaration
+        // block, then we add the error later on
+        if (send->block == nullptr) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // For all other cases, we have to re-write the block
+        // E.g.: `sig.abstract { void }` -> `sig { abstract.void }`
+        auto *block = ast::cast_tree<ast::Block>(send->block);
+        auto blockBody = ast::cast_tree<ast::Send>(block->body);
+
+        // If the blockBody is not a send, then we have a sequence of instructions inside the signature block
+        if (blockBody == nullptr) {
+            if (auto e = ctx.beginError(block->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Malformed signature: cannot have multiple instructions inside a signature block");
+            }
+
+            return tree;
+        }
+
+        return buildReplacement(ctx, send, blockBody);
+    }
+
+    bool firstReceiverIsSig(ast::Send *send) {
+        do {
+            if (send->fun == core::Names::sig()) {
+                this->previousSigSend = send;
+                return true;
+            }
+            send = ast::cast_tree<ast::Send>(send->recv);
+        } while (send);
+
+        return false;
+    }
+
+    void checkDuplicates(core::MutableContext ctx, ast::Send *sigBlock, core::NameRef fun) {
+        while (sigBlock) {
+            if (sigBlock->fun == fun) {
+                if (auto e = ctx.beginError(sigBlock->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Duplicate invocation of `{}` in signature declaration", fun.toString(ctx));
+                }
+            }
+
+            sigBlock = ast::cast_tree<ast::Send>(sigBlock->recv);
+        }
+    }
+
+    bool invalidChainedStatement(core::MutableContext ctx, ast::Send *send) {
+        if (send->fun == core::Names::params() || send->fun == core::Names::returns() ||
+            send->fun == core::Names::void_() || send->fun == core::Names::bind() ||
+            send->fun == core::Names::checked() || send->fun == core::Names::onFailure() ||
+            send->fun == core::Names::typeParameters()) {
+            auto receiver = ast::cast_tree<ast::Send>(send->recv);
+
+            while (receiver && receiver->fun != core::Names::sig()) {
+                receiver = ast::cast_tree<ast::Send>(receiver->recv);
+            }
```

Comment:
we're going to run this loop for literally every method call in the whole codebase

what do you think about first factoring out a check in `preTransformSend` that asks whether the current send has `sig` as one of its receivers, and early exiting from `preTransformSend` if not?

Reply:
it looks like you already have such a helper: `firstReceiverIsSig`. Can we just call that before calling invalidChainedStatement, and delete the loop from this method?